src/dbwrapper.h

```diff
@@ -97,6 +97,8 @@ class CDBBatch
 
     size_t size_estimate{0};
 
+    void WriteImpl(DataStream& ssKey, CDataStream& ssValue);
```

Comment:
> WriteImpl, ssKey and ssValue are all CDBBatch class members, I'm not sure passing these as parameters makes sense? And if so, I think ssKey should be const.

This gave me pause too, but it felt clearer to me this way. Let me know if you think passing them arguments here is more confusing.

> At first sight, I'm not entirely sure why ssKey and ssValue are CDBBatch members when they get resized and cleared in every operation, so perhaps cleaning that up is the better way to go, but I don't like the ambiguity the current implementation introduces.

AFAIU they don't get resized in every operation. Clear should not be de-allocating its memory and if it is already at the correct capacity, calling `reserve` should not be doing anything.

Reply:
Alright yeah I was a bit too quick with my initial suggestion, the current implementation makes sense from a (de)alloc perspective, thanks for pointing that both.

> Let me know if you think passing them arguments here is more confusing.

I still don't really like it, but I agree that taking these 2 arguments is the most natural interface for `WriteImpl`. The only reason we have this weirdness is just because `Write` wants to do some allocation optimization. So I'm okay keeping it like this.

Another approach that could work is to make `ssKey` and `ssValue` static vars? It means we would allocate one additional `DataStream` in `Erase`, but I think on the `CDBWrapper` level that's no big deal?

Reply:
Yeah, unless any objections this would have my preference, less side effects and easier to understand imo:

<details>
<summary>git diff on 26fe348e6e05d9b50614fdcba9cfb7a711ae86dc</summary>

```diff
diff --git a/src/dbwrapper.h b/src/dbwrapper.h
index 54c4685c48..48a6c9af00 100644
--- a/src/dbwrapper.h
+++ b/src/dbwrapper.h
@@ -92,9 +92,6 @@ private:
     const CDBWrapper &parent;
     leveldb::WriteBatch batch;
 
-    DataStream ssKey{};
-    CDataStream ssValue;
-
     size_t size_estimate{0};
 
     void WriteImpl(const DataStream& ssKey, CDataStream& ssValue);
@@ -103,7 +100,7 @@ public:
     /**
      * @param[in] _parent   CDBWrapper that this batch is to be submitted to
      */
-    explicit CDBBatch(const CDBWrapper& _parent) : parent(_parent), ssValue(SER_DISK, CLIENT_VERSION){};
+    explicit CDBBatch(const CDBWrapper& _parent) : parent(_parent){};
 
     void Clear()
     {
@@ -114,6 +111,9 @@ public:
     template <typename K, typename V>
     void Write(const K& key, const V& value)
     {
+        static DataStream ssKey{};
+        static CDataStream ssValue{SER_DISK, CLIENT_VERSION};
+
         ssKey.reserve(DBWRAPPER_PREALLOC_KEY_SIZE);
         ssValue.reserve(DBWRAPPER_PREALLOC_VALUE_SIZE);
         ssKey << key;
@@ -126,6 +126,8 @@ public:
     template <typename K>
     void Erase(const K& key)
     {
+        static DataStream ssKey{};
+
         ssKey.reserve(DBWRAPPER_PREALLOC_KEY_SIZE);
         ssKey << key;
         leveldb::Slice slKey(CharCast(ssKey.data()), ssKey.size());

```
</details>

Reply:
Not sure about making this a static var. Now the allocations survive beyond the lifetime of the `CDBBatch` objects and I am not sure if this is entirely safe.

Reply:
Good point. Okay, let's mark as resolved.